x-pack/plugin/apm-integration/src/main/java/org/elasticsearch/xpack/apm/APMTracer.java

```diff
@@ -75,8 +80,18 @@
     public APMTracer(Settings settings, ThreadPool threadPool, ClusterService clusterService) {
         this.endpoint = APM_ENDPOINT_SETTING.get(settings);
         this.token = APM_TOKEN_SETTING.get(settings);
+        this.enabled = APM_ENABLED_SETTING.get(settings);
         this.threadPool = Objects.requireNonNull(threadPool);
         this.clusterService = Objects.requireNonNull(clusterService);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(APM_ENABLED_SETTING, this::setEnabled);
+    }
+
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    private void setEnabled(boolean enabled) {
+        this.enabled = enabled;
```

Comment:
Could we completely shut everything down if this is set to false? Otherwise we have a bunch of threads and other things running in the background, which shouldn't really be doing anything but you never know...

Also should we eagerly discard any active spans?

Reply:
Right, we should probably be able to achieve this by calling the existing `doStart/doStop` methods on enabled/disabled respectively.

Reply:
I think we can't quite do that, `doStop()` blocks until the shutdown is complete but when disabling it dynamically we shouldn't wait.